<!doctype html><html lang=en class=no-js> <head><meta charset=utf-8><meta name=viewport content="width=device-width,initial-scale=1"><meta name=description content="The space will be used to provided documentation for khanlab NGS pipeline."><meta name=author content="Vineela Gangalapudi"><link href=usage/ rel=next><link rel=icon href=assets/images/favicon.png><meta name=generator content="mkdocs-1.5.3, mkdocs-material-9.5.14"><title>khanlab-NGS-Pipeline</title><link rel=stylesheet href=assets/stylesheets/main.10ba22f1.min.css><link rel=stylesheet href=assets/stylesheets/palette.06af60db.min.css><link rel=preconnect href=https://fonts.gstatic.com crossorigin><link rel=stylesheet href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback"><style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style><link rel=stylesheet href=overrides/extra.css><script>__md_scope=new URL(".",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script></head> <body dir=ltr data-md-color-scheme=default data-md-color-primary=indigo data-md-color-accent=indigo> <input class=md-toggle data-md-toggle=drawer type=checkbox id=__drawer autocomplete=off> <input class=md-toggle data-md-toggle=search type=checkbox id=__search autocomplete=off> <label class=md-overlay for=__drawer></label> <div data-md-component=skip> <a href=#introduction class=md-skip> Skip to content </a> </div> <div data-md-component=announce> </div> <header class="md-header md-header--shadow" data-md-component=header> <nav class="md-header__inner md-grid" aria-label=Header> <a href=. title=khanlab-NGS-Pipeline class="md-header__button md-logo" aria-label=khanlab-NGS-Pipeline data-md-component=logo> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54Z"/></svg> </a> <label class="md-header__button md-icon" for=__drawer> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg> </label> <div class=md-header__title data-md-component=header-title> <div class=md-header__ellipsis> <div class=md-header__topic> <span class=md-ellipsis> khanlab-NGS-Pipeline </span> </div> <div class=md-header__topic data-md-component=header-topic> <span class=md-ellipsis> Introduction </span> </div> </div> </div> <form class=md-header__option data-md-component=palette> <input class=md-option data-md-color-media="(prefers-color-scheme: light)" data-md-color-scheme=default data-md-color-primary=indigo data-md-color-accent=indigo aria-label="Switch to dark mode" type=radio name=__palette id=__palette_0> <label class="md-header__button md-icon" title="Switch to dark mode" for=__palette_1 hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M12 8a4 4 0 0 0-4 4 4 4 0 0 0 4 4 4 4 0 0 0 4-4 4 4 0 0 0-4-4m0 10a6 6 0 0 1-6-6 6 6 0 0 1 6-6 6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12 20 8.69Z"/></svg> </label> <input class=md-option data-md-color-media="(prefers-color-scheme: dark)" data-md-color-scheme=slate data-md-color-primary=indigo data-md-color-accent=indigo aria-label="Switch to light mode" type=radio name=__palette id=__palette_1> <label class="md-header__button md-icon" title="Switch to light mode" for=__palette_0 hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M12 18c-.89 0-1.74-.2-2.5-.55C11.56 16.5 13 14.42 13 12c0-2.42-1.44-4.5-3.5-5.45C10.26 6.2 11.11 6 12 6a6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12 20 8.69Z"/></svg> </label> </form> <script>var media,input,key,value,palette=__md_get("__palette");if(palette&&palette.color){"(prefers-color-scheme)"===palette.color.media&&(media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']"),palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent"));for([key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script> <label class="md-header__button md-icon" for=__search> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg> </label> <div class=md-search data-md-component=search role=dialog> <label class=md-search__overlay for=__search></label> <div class=md-search__inner role=search> <form class=md-search__form name=search> <input type=text class=md-search__input name=query aria-label=Search placeholder=Search autocapitalize=off autocorrect=off autocomplete=off spellcheck=false data-md-component=search-query required> <label class="md-search__icon md-icon" for=__search> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg> </label> <nav class=md-search__options aria-label=Search> <button type=reset class="md-search__icon md-icon" title=Clear aria-label=Clear tabindex=-1> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg> </button> </nav> </form> <div class=md-search__output> <div class=md-search__scrollwrap data-md-scrollfix> <div class=md-search-result data-md-component=search-result> <div class=md-search-result__meta> Initializing search </div> <ol class=md-search-result__list role=presentation></ol> </div> </div> </div> </div> </div> <div class=md-header__source> <a href=https://github.com/CCRGeneticsBranch/Oncogenomics_NF_WF title="Go to repository" class=md-source data-md-component=source> <div class="md-source__icon md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 496 512"><!-- Font Awesome Free 6.5.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M165.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6zm-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3zm44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9zM244.8 8C106.1 8 0 113.3 0 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C428.2 457.8 496 362.9 496 252 496 113.3 383.5 8 244.8 8zM97.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1zm-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7zm32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1zm-11.4-14.7c-1.6 1-1.6 3.6 0 5.9 1.6 2.3 4.3 3.3 5.6 2.3 1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2z"/></svg> </div> <div class=md-source__repository> AWS_POC_MVP_NF </div> </a> </div> </nav> </header> <div class=md-container data-md-component=container> <main class=md-main data-md-component=main> <div class="md-main__inner md-grid"> <div class="md-sidebar md-sidebar--primary" data-md-component=sidebar data-md-type=navigation> <div class=md-sidebar__scrollwrap> <div class=md-sidebar__inner> <nav class="md-nav md-nav--primary" aria-label=Navigation data-md-level=0> <label class=md-nav__title for=__drawer> <a href=. title=khanlab-NGS-Pipeline class="md-nav__button md-logo" aria-label=khanlab-NGS-Pipeline data-md-component=logo> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54Z"/></svg> </a> khanlab-NGS-Pipeline </label> <div class=md-nav__source> <a href=https://github.com/CCRGeneticsBranch/Oncogenomics_NF_WF title="Go to repository" class=md-source data-md-component=source> <div class="md-source__icon md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 496 512"><!-- Font Awesome Free 6.5.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M165.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6zm-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3zm44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9zM244.8 8C106.1 8 0 113.3 0 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C428.2 457.8 496 362.9 496 252 496 113.3 383.5 8 244.8 8zM97.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1zm-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7zm32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1zm-11.4-14.7c-1.6 1-1.6 3.6 0 5.9 1.6 2.3 4.3 3.3 5.6 2.3 1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2z"/></svg> </div> <div class=md-source__repository> AWS_POC_MVP_NF </div> </a> </div> <ul class=md-nav__list data-md-scrollfix> <li class="md-nav__item md-nav__item--active"> <input class="md-nav__toggle md-toggle" type=checkbox id=__toc> <label class="md-nav__link md-nav__link--active" for=__toc> <span class=md-ellipsis> Introduction </span> <span class="md-nav__icon md-icon"></span> </label> <a href=. class="md-nav__link md-nav__link--active"> <span class=md-ellipsis> Introduction </span> </a> <nav class="md-nav md-nav--secondary" aria-label="Table of contents"> <label class=md-nav__title for=__toc> <span class="md-nav__icon md-icon"></span> Table of contents </label> <ul class=md-nav__list data-md-component=toc data-md-scrollfix> <li class=md-nav__item> <a href=#introduction class=md-nav__link> <span class=md-ellipsis> Introduction </span> </a> <nav class=md-nav aria-label=Introduction> <ul class=md-nav__list> <li class=md-nav__item> <a href=#who-this-is-for class=md-nav__link> <span class=md-ellipsis> Who This Is For </span> </a> </li> <li class=md-nav__item> <a href=#rnaseq-workflow-dag class=md-nav__link> <span class=md-ellipsis> RNAseq Workflow DAG </span> </a> </li> <li class=md-nav__item> <a href=#exome-workflow-dag class=md-nav__link> <span class=md-ellipsis> Exome Workflow DAG </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#workflow-tools class=md-nav__link> <span class=md-ellipsis> Workflow Tools </span> </a> </li> <li class=md-nav__item> <a href=#prerequisites class=md-nav__link> <span class=md-ellipsis> Prerequisites </span> </a> </li> <li class=md-nav__item> <a href=#architecture class=md-nav__link> <span class=md-ellipsis> Architecture </span> </a> <nav class=md-nav aria-label=Architecture> <ul class=md-nav__list> <li class=md-nav__item> <a href=#sequencing-capture-kits class=md-nav__link> <span class=md-ellipsis> Sequencing capture kits </span> </a> </li> </ul> </nav> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=usage/ class=md-nav__link> <span class=md-ellipsis> Usage </span> </a> </li> <li class=md-nav__item> <a href=pipeline_launch/ class=md-nav__link> <span class=md-ellipsis> Running the Pipeline </span> </a> </li> <li class=md-nav__item> <a href=results/ class=md-nav__link> <span class=md-ellipsis> Results </span> </a> </li> <li class=md-nav__item> <a href=references/ class=md-nav__link> <span class=md-ellipsis> Pipeline References & Tools </span> </a> </li> <li class=md-nav__item> <a href=dockers/ class=md-nav__link> <span class=md-ellipsis> Dockers </span> </a> </li> </ul> </nav> </div> </div> </div> <div class="md-sidebar md-sidebar--secondary" data-md-component=sidebar data-md-type=toc> <div class=md-sidebar__scrollwrap> <div class=md-sidebar__inner> <nav class="md-nav md-nav--secondary" aria-label="Table of contents"> <label class=md-nav__title for=__toc> <span class="md-nav__icon md-icon"></span> Table of contents </label> <ul class=md-nav__list data-md-component=toc data-md-scrollfix> <li class=md-nav__item> <a href=#introduction class=md-nav__link> <span class=md-ellipsis> Introduction </span> </a> <nav class=md-nav aria-label=Introduction> <ul class=md-nav__list> <li class=md-nav__item> <a href=#who-this-is-for class=md-nav__link> <span class=md-ellipsis> Who This Is For </span> </a> </li> <li class=md-nav__item> <a href=#rnaseq-workflow-dag class=md-nav__link> <span class=md-ellipsis> RNAseq Workflow DAG </span> </a> </li> <li class=md-nav__item> <a href=#exome-workflow-dag class=md-nav__link> <span class=md-ellipsis> Exome Workflow DAG </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#workflow-tools class=md-nav__link> <span class=md-ellipsis> Workflow Tools </span> </a> </li> <li class=md-nav__item> <a href=#prerequisites class=md-nav__link> <span class=md-ellipsis> Prerequisites </span> </a> </li> <li class=md-nav__item> <a href=#architecture class=md-nav__link> <span class=md-ellipsis> Architecture </span> </a> <nav class=md-nav aria-label=Architecture> <ul class=md-nav__list> <li class=md-nav__item> <a href=#sequencing-capture-kits class=md-nav__link> <span class=md-ellipsis> Sequencing capture kits </span> </a> </li> </ul> </nav> </li> </ul> </nav> </div> </div> </div> <div class=md-content data-md-component=content> <article class="md-content__inner md-typeset"> <h1>Introduction</h1> <h2 id=introduction>Introduction<a class=headerlink href=#introduction title="Permanent link">&para;</a></h2> <p>This bioinformatics pipeline is containerized and platform-independent, built using a Nextflow DSL2 workflow, and deployed on both the NIH <strong>Biowulf HPC cluster</strong> and the <strong>AWS cloud</strong>. </p> <p>This <strong>Exome–RNA-seq workflow</strong> accepts a samplesheet and FASTQ files as input, then performs a comprehensive set of analyses, including: </p> <ul> <li>Extensive quality control </li> <li>Mutation calling </li> <li>Tumor mutational burden assessment </li> <li>Mutational signature analysis </li> <li>HLA typing </li> <li>Copy-number variation detection </li> <li>T-cell infiltration prediction </li> <li>Neoantigen prediction </li> <li>Gene expression profiling </li> <li>Fusion detection </li> <li>Variant calling and annotation </li> </ul> <p>This end-to-end suite supports deep genomic characterization for both research and institutional datasets at NIH. </p> <p>Results can be explored via the <a href=https://oncogenomics.ccr.cancer.gov/production/public/ >ClinOmics Data Portal</a>, which provides a secure, interactive, user friendly web interface for data visualization and interpretation. It is well suited for both individual patient level and cohort wide cancer genomics analysis.</p> <h3 id=who-this-is-for>Who This Is For<a class=headerlink href=#who-this-is-for title="Permanent link">&para;</a></h3> <p>This workflow is intended for researchers and bioinformaticians working with exome and RNA-seq datasets who need a reproducible, scalable, and validated solution for comprehensive genomic analysis. It is designed for use in both <strong>HPC</strong> and <strong>cloud computing</strong> environments. </p> <p>Here is a Snapshot of our RNAseq and Exome workflows.</p> <h3 id=rnaseq-workflow-dag>RNAseq Workflow DAG<a class=headerlink href=#rnaseq-workflow-dag title="Permanent link">&para;</a></h3> <p><img alt="RNAseq Workflow DAG" src=RNAseq_DAG.png></p> <h3 id=exome-workflow-dag>Exome Workflow DAG<a class=headerlink href=#exome-workflow-dag title="Permanent link">&para;</a></h3> <p><img alt=Exome_workflow src=Exome_DAG.png></p> <h2 id=workflow-tools>Workflow Tools<a class=headerlink href=#workflow-tools title="Permanent link">&para;</a></h2> <div class="tabbed-set tabbed-alternate" data-tabs=1:3><input checked=checked id=__tabbed_1_1 name=__tabbed_1 type=radio><input id=__tabbed_1_2 name=__tabbed_1 type=radio><input id=__tabbed_1_3 name=__tabbed_1 type=radio><div class=tabbed-labels><label for=__tabbed_1_1>RNA-seq Tools</label><label for=__tabbed_1_2>Exome Tools</label><label for=__tabbed_1_3>Common Tools</label></div> <div class=tabbed-content> <div class=tabbed-block> <table> <thead> <tr> <th>Stage</th> <th>Tool(s)</th> <th>Purpose</th> </tr> </thead> <tbody> <tr> <td>Alignment</td> <td>STAR</td> <td>Align RNA-seq reads to the reference genome with splice-aware mapping</td> </tr> <tr> <td>Strandedness Assessment</td> <td>ngsderive</td> <td>Determine library strandedness from aligned BAM files</td> </tr> <tr> <td>Expression Quantification</td> <td>RSEM</td> <td>Quantify gene and transcript abundance</td> </tr> <tr> <td>Post-mapping QC</td> <td>Picard CollectRNASeqMetrics, RNA-SeQC</td> <td>Evaluate mapping quality, insert size, and gene body coverage</td> </tr> <tr> <td>BAM Processing</td> <td>GATK SplitNCigarReads, RealignerTargetCreator, IndelRealigner, BaseRecalibrator, PrintReads</td> <td>Prepare RNA-seq BAMs for variant calling by handling spliced reads, realigning indels, and recalibrating base quality scores</td> </tr> <tr> <td>Variant Calling</td> <td>GATK HaplotypeCaller</td> <td>Identify germline SNPs and indels from RNA-seq data</td> </tr> <tr> <td>Fusion Detection</td> <td>STAR-Fusion, Arriba, FusionCatcher</td> <td>Detect gene fusions; integrate consensus calls across multiple tools</td> </tr> <tr> <td>HLA Calling</td> <td>HLA-HD, OptiType</td> <td>Predict HLA types from RNA-seq data</td> </tr> <tr> <td>Variant Annotation</td> <td>ANNOVAR</td> <td>Annotate variants with functional consequences and external database references</td> </tr> </tbody> </table> </div> <div class=tabbed-block> <table> <thead> <tr> <th>Stage</th> <th>Tool(s)</th> <th>Purpose</th> </tr> </thead> <tbody> <tr> <td>Alignment</td> <td>BWA-MEM</td> <td>Align exome reads to the reference genome</td> </tr> <tr> <td>BAM QC</td> <td>Picard CollectAlignmentSummaryMetrics, Picard CollectInsertSizeMetrics</td> <td>Evaluate mapping quality, duplication rates, and insert size</td> </tr> <tr> <td>BAM Processing</td> <td>GATK MarkDuplicates, BaseRecalibrator, ApplyBQSR</td> <td>Mark duplicates, recalibrate base quality scores, and prepare BAMs for variant calling</td> </tr> <tr> <td>Germline Variant Calling</td> <td>GATK HaplotypeCaller</td> <td>Identify germline SNVs and indels</td> </tr> <tr> <td>Somatic Variant Calling</td> <td>Manta, Strelka2, GATK Mutect2</td> <td>Identify somatic SNVs and indels</td> </tr> <tr> <td>Copy Number Calling</td> <td>CNVkit</td> <td>Detect copy number alterations</td> </tr> <tr> <td>Variant Annotation</td> <td>ANNOVAR, SnpEff, VEP</td> <td>Annotate variants with functional consequences and external database references</td> </tr> <tr> <td>T-cell Infiltration Estimation</td> <td>TcellExTRECT</td> <td>Estimate tumor-infiltrating T-cell abundance from sequencing data</td> </tr> <tr> <td>Tumor Ploidy &amp; Cellularity</td> <td>Sequenza</td> <td>Infer tumor purity and ploidy</td> </tr> <tr> <td>Microsatellite Instability</td> <td>MANTIS</td> <td>Detect microsatellite instability from tumor/normal pairs</td> </tr> <tr> <td>Mutational Signatures</td> <td>SigProfilerExtractor, COSMIC Signature Database</td> <td>Identify mutational signatures present in the tumor genome</td> </tr> <tr> <td>HLA Typing</td> <td>HLA-HD, OptiType</td> <td>Predict HLA types from sequencing data</td> </tr> <tr> <td>Neoantigen Prediction</td> <td>pVACseq</td> <td>Predict candidate neoantigens from tumor-specific variants</td> </tr> <tr> <td>Concordance Analysis</td> <td>Conpair</td> <td>Verify tumor/normal sample pairing via genotype concordance</td> </tr> </tbody> </table> </div> <div class=tabbed-block> <table> <thead> <tr> <th>Stage</th> <th>Tool(s)</th> <th>Purpose</th> </tr> </thead> <tbody> <tr> <td>Quality Control</td> <td>FastQC, MultiQC</td> <td>Assess raw read quality and generate aggregated QC reports</td> </tr> <tr> <td>Contamination Screening</td> <td>Kraken2, Krona, FastQ Screen</td> <td>Identify potential contamination from non-target species</td> </tr> <tr> <td>Read Trimming</td> <td>Cutadapt</td> <td>Remove adapters, low-quality bases, and trim reads for improved alignment</td> </tr> <tr> <td>BAM QC &amp; Coverage Analysis</td> <td>samtools, bamUtils, mpileup, bedtools</td> <td>Generate hotspot coverage plots, transcript coverage plots, circos plots, and perform BAM compression/clean-up</td> </tr> <tr> <td>QC Aggregation</td> <td>MultiQC</td> <td>Aggregate QC metrics across all steps into a single interactive report</td> </tr> </tbody> </table> </div> </div> </div> <div class="admonition tip"> <p class=admonition-title>Before You Begin</p> <p>If you are setting up this workflow for the first time, ensure your environment meets all prerequisites and dependencies.<br> See the <a href=#prerequisites>Prerequisites</a> section below and If you are new to Nextflow, please refer to <a href=https://www.nextflow.io/docs/latest/ >this page</a> for quick introduction.</p> </div> <h2 id=prerequisites>Prerequisites<a class=headerlink href=#prerequisites title="Permanent link">&para;</a></h2> <p>To run this workflow, you will need the following software:</p> <div class=highlight><pre><span></span><code>    Nextflow &gt;= 21.04.3
    Singularity 3.10.5
    Graphviz 2.40
</code></pre></div> <h2 id=architecture>Architecture<a class=headerlink href=#architecture title="Permanent link">&para;</a></h2> <ol> <li>The workflow is located in the Khanlab space: <code>/data/khanlab/projects/Nextflow_dev/dev/AWS_POC_MVP_NF</code>.</li> <li> <p>The <strong><code>biowulf_nextflow.config</code></strong> file in this directory contains the master configuration. A <code>config/</code> subfolder contains additional configuration files:</p> <div class=highlight><pre><span></span><code>config/
├── aws_cluster_test.config
├── aws_ec2_test.config
├── aws_params.config
├── biowulf_cluster_test.config
├── biowulf_s3_test_params.config
├── biowulf_singularity.config
├── biowulf_small_test_params.config
├── biowulf_small_test_params_hg38.config
├── docker.config
├── GRCm39.config
└── omics.config
</code></pre></div> </li> <li> <p>Pipeline can be launched using the script launch.sh. It takes samplesheet.csv as a mandatory input. </p> </li> <li>Guidelines to create an input samplesheet can be found <a href=usage/ >here</a>.</li> <li>All the references, annotation and bed files are currently located under <code>/data/khanlab</code>. We currently support data processing for these capture kits.</li> </ol> <h3 id=sequencing-capture-kits>Sequencing capture kits<a class=headerlink href=#sequencing-capture-kits title="Permanent link">&para;</a></h3> <table> <thead> <tr> <th>RNAseq</th> <th>Exome</th> </tr> </thead> <tbody> <tr> <td>Access</td> <td>clin_ex_v1</td> </tr> <tr> <td>PolyA</td> <td>seqcapez.hu.ex.v3</td> </tr> <tr> <td>PolyA_stranded</td> <td>Agilent_v7</td> </tr> <tr> <td>Ribozero</td> <td>idt_v2_plus</td> </tr> <tr> <td>smartrna</td> <td>xgen_hyb_panelv2</td> </tr> <tr> <td></td> <td>seqcapez_hu_ex_v3</td> </tr> </tbody> </table> <p>If you want to process the data sequenced by other kits, please reach out to <a href=mailto:vineela.gangalapudi@nih.gov>Vineela Gangalapudi</a>.</p> <aside class=md-source-file> <span class=md-source-file__fact> <span class=md-icon title="Last update"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M21 13.1c-.1 0-.3.1-.4.2l-1 1 2.1 2.1 1-1c.2-.2.2-.6 0-.8l-1.3-1.3c-.1-.1-.2-.2-.4-.2m-1.9 1.8-6.1 6V23h2.1l6.1-6.1-2.1-2M12.5 7v5.2l4 2.4-1 1L11 13V7h1.5M11 21.9c-5.1-.5-9-4.8-9-9.9C2 6.5 6.5 2 12 2c5.3 0 9.6 4.1 10 9.3-.3-.1-.6-.2-1-.2s-.7.1-1 .2C19.6 7.2 16.2 4 12 4c-4.4 0-8 3.6-8 8 0 4.1 3.1 7.5 7.1 7.9l-.1.2v1.8Z"/></svg> </span> 2025-08-13 </span> </aside> </article> </div> <script>var tabs=__md_get("__tabs");if(Array.isArray(tabs))e:for(var set of document.querySelectorAll(".tabbed-set")){var tab,labels=set.querySelector(".tabbed-labels");for(tab of tabs)for(var label of labels.getElementsByTagName("label"))if(label.innerText.trim()===tab){var input=document.getElementById(label.htmlFor);input.checked=!0;continue e}}</script> <script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script> </div> </main> <footer class=md-footer> <div class="md-footer-meta md-typeset"> <div class="md-footer-meta__inner md-grid"> <div class=md-copyright> <div class=md-copyright__highlight> Copyright &copy; 2025 CCR Genetics Branch </div> Made with <a href=https://squidfunk.github.io/mkdocs-material/ target=_blank rel=noopener> Material for MkDocs </a> </div> </div> </div> </footer> </div> <div class=md-dialog data-md-component=dialog> <div class="md-dialog__inner md-typeset"></div> </div> <script id=__config type=application/json>{"base": ".", "features": ["content.code.copy", "content.code.annotate", "content.tabs.link"], "search": "assets/javascripts/workers/search.b8dbb3d2.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script> <script src=assets/javascripts/bundle.bd41221c.min.js></script> </body> </html>