<!doctype html><html lang=en class=no-js> <head><meta charset=utf-8><meta name=viewport content="width=device-width,initial-scale=1"><meta name=description content="The space will be used to provided documentation for khanlab NGS pipeline."><meta name=author content="Vineela Gangalapudi"><link rel=prev href=..><link href=../pipeline_launch/ rel=next><link rel=icon href=../assets/images/favicon.png><meta name=generator content="mkdocs-1.5.3, mkdocs-material-9.5.14"><title>Usage - khanlab-NGS-Pipeline</title><link rel=stylesheet href=../assets/stylesheets/main.10ba22f1.min.css><link rel=stylesheet href=../assets/stylesheets/palette.06af60db.min.css><link rel=preconnect href=https://fonts.gstatic.com crossorigin><link rel=stylesheet href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback"><style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style><link rel=stylesheet href=../overrides/extra.css><script>__md_scope=new URL("..",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script></head> <body dir=ltr data-md-color-scheme=default data-md-color-primary=indigo data-md-color-accent=indigo> <input class=md-toggle data-md-toggle=drawer type=checkbox id=__drawer autocomplete=off> <input class=md-toggle data-md-toggle=search type=checkbox id=__search autocomplete=off> <label class=md-overlay for=__drawer></label> <div data-md-component=skip> <a href=#usage class=md-skip> Skip to content </a> </div> <div data-md-component=announce> </div> <header class="md-header md-header--shadow" data-md-component=header> <nav class="md-header__inner md-grid" aria-label=Header> <a href=.. title=khanlab-NGS-Pipeline class="md-header__button md-logo" aria-label=khanlab-NGS-Pipeline data-md-component=logo> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54Z"/></svg> </a> <label class="md-header__button md-icon" for=__drawer> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg> </label> <div class=md-header__title data-md-component=header-title> <div class=md-header__ellipsis> <div class=md-header__topic> <span class=md-ellipsis> khanlab-NGS-Pipeline </span> </div> <div class=md-header__topic data-md-component=header-topic> <span class=md-ellipsis> Usage </span> </div> </div> </div> <form class=md-header__option data-md-component=palette> <input class=md-option data-md-color-media="(prefers-color-scheme: light)" data-md-color-scheme=default data-md-color-primary=indigo data-md-color-accent=indigo aria-label="Switch to dark mode" type=radio name=__palette id=__palette_0> <label class="md-header__button md-icon" title="Switch to dark mode" for=__palette_1 hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M12 8a4 4 0 0 0-4 4 4 4 0 0 0 4 4 4 4 0 0 0 4-4 4 4 0 0 0-4-4m0 10a6 6 0 0 1-6-6 6 6 0 0 1 6-6 6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12 20 8.69Z"/></svg> </label> <input class=md-option data-md-color-media="(prefers-color-scheme: dark)" data-md-color-scheme=slate data-md-color-primary=indigo data-md-color-accent=indigo aria-label="Switch to light mode" type=radio name=__palette id=__palette_1> <label class="md-header__button md-icon" title="Switch to light mode" for=__palette_0 hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M12 18c-.89 0-1.74-.2-2.5-.55C11.56 16.5 13 14.42 13 12c0-2.42-1.44-4.5-3.5-5.45C10.26 6.2 11.11 6 12 6a6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12 20 8.69Z"/></svg> </label> </form> <script>var media,input,key,value,palette=__md_get("__palette");if(palette&&palette.color){"(prefers-color-scheme)"===palette.color.media&&(media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']"),palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent"));for([key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script> <label class="md-header__button md-icon" for=__search> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg> </label> <div class=md-search data-md-component=search role=dialog> <label class=md-search__overlay for=__search></label> <div class=md-search__inner role=search> <form class=md-search__form name=search> <input type=text class=md-search__input name=query aria-label=Search placeholder=Search autocapitalize=off autocorrect=off autocomplete=off spellcheck=false data-md-component=search-query required> <label class="md-search__icon md-icon" for=__search> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg> </label> <nav class=md-search__options aria-label=Search> <button type=reset class="md-search__icon md-icon" title=Clear aria-label=Clear tabindex=-1> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg> </button> </nav> </form> <div class=md-search__output> <div class=md-search__scrollwrap data-md-scrollfix> <div class=md-search-result data-md-component=search-result> <div class=md-search-result__meta> Initializing search </div> <ol class=md-search-result__list role=presentation></ol> </div> </div> </div> </div> </div> <div class=md-header__source> <a href=https://github.com/CCRGeneticsBranch/Oncogenomics_NF_WF title="Go to repository" class=md-source data-md-component=source> <div class="md-source__icon md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 496 512"><!-- Font Awesome Free 6.5.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M165.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6zm-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3zm44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9zM244.8 8C106.1 8 0 113.3 0 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C428.2 457.8 496 362.9 496 252 496 113.3 383.5 8 244.8 8zM97.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1zm-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7zm32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1zm-11.4-14.7c-1.6 1-1.6 3.6 0 5.9 1.6 2.3 4.3 3.3 5.6 2.3 1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2z"/></svg> </div> <div class=md-source__repository> AWS_POC_MVP_NF </div> </a> </div> </nav> </header> <div class=md-container data-md-component=container> <main class=md-main data-md-component=main> <div class="md-main__inner md-grid"> <div class="md-sidebar md-sidebar--primary" data-md-component=sidebar data-md-type=navigation> <div class=md-sidebar__scrollwrap> <div class=md-sidebar__inner> <nav class="md-nav md-nav--primary" aria-label=Navigation data-md-level=0> <label class=md-nav__title for=__drawer> <a href=.. title=khanlab-NGS-Pipeline class="md-nav__button md-logo" aria-label=khanlab-NGS-Pipeline data-md-component=logo> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54Z"/></svg> </a> khanlab-NGS-Pipeline </label> <div class=md-nav__source> <a href=https://github.com/CCRGeneticsBranch/Oncogenomics_NF_WF title="Go to repository" class=md-source data-md-component=source> <div class="md-source__icon md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 496 512"><!-- Font Awesome Free 6.5.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M165.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6zm-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3zm44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9zM244.8 8C106.1 8 0 113.3 0 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C428.2 457.8 496 362.9 496 252 496 113.3 383.5 8 244.8 8zM97.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1zm-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7zm32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1zm-11.4-14.7c-1.6 1-1.6 3.6 0 5.9 1.6 2.3 4.3 3.3 5.6 2.3 1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2z"/></svg> </div> <div class=md-source__repository> AWS_POC_MVP_NF </div> </a> </div> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=.. class=md-nav__link> <span class=md-ellipsis> Introduction </span> </a> </li> <li class="md-nav__item md-nav__item--active"> <input class="md-nav__toggle md-toggle" type=checkbox id=__toc> <label class="md-nav__link md-nav__link--active" for=__toc> <span class=md-ellipsis> Usage </span> <span class="md-nav__icon md-icon"></span> </label> <a href=./ class="md-nav__link md-nav__link--active"> <span class=md-ellipsis> Usage </span> </a> <nav class="md-nav md-nav--secondary" aria-label="Table of contents"> <label class=md-nav__title for=__toc> <span class="md-nav__icon md-icon"></span> Table of contents </label> <ul class=md-nav__list data-md-component=toc data-md-scrollfix> <li class=md-nav__item> <a href=#usage class=md-nav__link> <span class=md-ellipsis> Usage </span> </a> </li> <li class=md-nav__item> <a href=#running-on-biowulf class=md-nav__link> <span class=md-ellipsis> Running on Biowulf </span> </a> <nav class=md-nav aria-label="Running on Biowulf"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#step-1-copy-the-launch-script class=md-nav__link> <span class=md-ellipsis> Step 1 — Copy the launch script </span> </a> </li> <li class=md-nav__item> <a href=#step-2-view-usage-instructions class=md-nav__link> <span class=md-ellipsis> Step 2 — View usage instructions </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#preparing-the-samplesheet class=md-nav__link> <span class=md-ellipsis> Preparing the Samplesheet </span> </a> <nav class=md-nav aria-label="Preparing the Samplesheet"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#1-build-from-a-master-sheet-recommended class=md-nav__link> <span class=md-ellipsis> 1. Build from a Master Sheet (Recommended) </span> </a> </li> <li class=md-nav__item> <a href=#2-build-your-own-samplesheet class=md-nav__link> <span class=md-ellipsis> 2. Build Your Own Samplesheet </span> </a> </li> </ul> </nav> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=../pipeline_launch/ class=md-nav__link> <span class=md-ellipsis> Running the Pipeline </span> </a> </li> <li class=md-nav__item> <a href=../references/ class=md-nav__link> <span class=md-ellipsis> Pipeline References & Tools </span> </a> </li> <li class=md-nav__item> <a href=../dockers/ class=md-nav__link> <span class=md-ellipsis> Dockers </span> </a> </li> <li class=md-nav__item> <a href=../results/ class=md-nav__link> <span class=md-ellipsis> Results </span> </a> </li> </ul> </nav> </div> </div> </div> <div class="md-sidebar md-sidebar--secondary" data-md-component=sidebar data-md-type=toc> <div class=md-sidebar__scrollwrap> <div class=md-sidebar__inner> <nav class="md-nav md-nav--secondary" aria-label="Table of contents"> <label class=md-nav__title for=__toc> <span class="md-nav__icon md-icon"></span> Table of contents </label> <ul class=md-nav__list data-md-component=toc data-md-scrollfix> <li class=md-nav__item> <a href=#usage class=md-nav__link> <span class=md-ellipsis> Usage </span> </a> </li> <li class=md-nav__item> <a href=#running-on-biowulf class=md-nav__link> <span class=md-ellipsis> Running on Biowulf </span> </a> <nav class=md-nav aria-label="Running on Biowulf"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#step-1-copy-the-launch-script class=md-nav__link> <span class=md-ellipsis> Step 1 — Copy the launch script </span> </a> </li> <li class=md-nav__item> <a href=#step-2-view-usage-instructions class=md-nav__link> <span class=md-ellipsis> Step 2 — View usage instructions </span> </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#preparing-the-samplesheet class=md-nav__link> <span class=md-ellipsis> Preparing the Samplesheet </span> </a> <nav class=md-nav aria-label="Preparing the Samplesheet"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#1-build-from-a-master-sheet-recommended class=md-nav__link> <span class=md-ellipsis> 1. Build from a Master Sheet (Recommended) </span> </a> </li> <li class=md-nav__item> <a href=#2-build-your-own-samplesheet class=md-nav__link> <span class=md-ellipsis> 2. Build Your Own Samplesheet </span> </a> </li> </ul> </nav> </li> </ul> </nav> </div> </div> </div> <div class=md-content data-md-component=content> <article class="md-content__inner md-typeset"> <h1>Usage</h1> <h2 id=usage>Usage<a class=headerlink href=#usage title="Permanent link">&para;</a></h2> <p>This page describes how to run the workflow on <strong>Biowulf</strong>.<br> For installation and prerequisites, see the <a href=../ >Installation</a> page.</p> <hr> <h2 id=running-on-biowulf>Running on Biowulf<a class=headerlink href=#running-on-biowulf title="Permanent link">&para;</a></h2> <p>On Biowulf, the pipeline is hosted in the Khanlab space.<br> A helper script, <code>launch.sh</code>, is provided to simplify execution. This script wraps the <code>nextflow run</code> command with Biowulf-specific settings and automatically configures output directories, logging, and profiles.</p> <h3 id=step-1-copy-the-launch-script>Step 1 — Copy the launch script<a class=headerlink href=#step-1-copy-the-launch-script title="Permanent link">&para;</a></h3> <details> <summary>launch.sh</summary> <div class=highlight><pre><span></span><code>#!/bin/bash

# Set default values
DEFAULT_OUTDIR=&quot;/data/khanlab/projects/processed_DATA&quot;
DEFAULT_GENOME=&quot;hg19&quot;
DEFAULT_PLATFORM=&quot;biowulf&quot;

# Check if the required samplesheet argument is provided
if [[ &quot;$#&quot; -lt 1 ]]; then
    echo &quot;Usage: $0 &lt;samplesheet_with_full_path&gt; [output_directory] [genome]&quot;
    echo &quot;This script requires at least one positional argument:&quot;
    echo &quot;1. Path to samplesheet&quot;
    echo &quot;Optional arguments:&quot;
    echo &quot;2. Path to results directory (default: $DEFAULT_OUTDIR)&quot;
    echo &quot;3. Genome name. Accepted values are hg19 and mm39 (default: $DEFAULT_GENOME)&quot;
    exit 1
fi

# Assign the first argument (samplesheet) to a variable
export SAMPLESHEET=$1

# Assign the second argument (output_directory) if provided, otherwise use the default
export OUTDIR=${2:-$DEFAULT_OUTDIR}

# Assign the third argument (genome) if provided, otherwise use the default
export GENOME=${3:-$DEFAULT_GENOME}

export PLATFORM=${4:-$DEFAULT_PLATFORM}

# Ensure the genome name is valid
if [[ &quot;$GENOME&quot; != &quot;hg19&quot; &amp;&amp; &quot;$GENOME&quot; != &quot;mm39&quot; ]]; then
    echo &quot;Invalid genome specified. Accepted values are hg19 and mm39.&quot;
    exit 1
fi

WF_HOME=&quot;/data/khanlab/projects/Nextflow_dev/dev/AWS_POC_MVP_NF&quot;
CONFIG_FILE=&quot;$WF_HOME/biowulf_nextflow.config&quot;

#export PATIENT=$(awk -F&#39;,&#39; &#39;NR==1 {for (i=1; i&lt;=NF; i++) if ($i==&quot;sample&quot;) s=i} NR&gt;1 {print $s}&#39; &quot;$SAMPLESHEET&quot; | sort | uniq)
#export CASENAME=$(awk -F&#39;,&#39; &#39;NR==1 {for (i=1; i&lt;=NF; i++) if ($i==&quot;casename&quot;) c=i} NR&gt;1 {print $c}&#39; &quot;$SAMPLESHEET&quot; | sort | uniq)
export PATIENT=$(python3 -c &#39;import csv,sys; r=csv.DictReader(open(sys.argv[1])); print(&quot;\n&quot;.join(sorted(set(row[&quot;sample&quot;] for row in r))))&#39; &quot;$SAMPLESHEET&quot;)
export CASENAME=$(python3 -c &#39;import csv,sys; r=csv.DictReader(open(sys.argv[1])); print(&quot;\n&quot;.join(sorted(set(row[&quot;casename&quot;] for row in r))))&#39; &quot;$SAMPLESHEET&quot;)

export RESULTSDIR=&quot;$OUTDIR/$PATIENT/$CASENAME&quot;

mkdir -p &quot;$RESULTSDIR&quot;

export LOG=&quot;$RESULTSDIR/log&quot;

mkdir -p &quot;$LOG&quot;

export NXF_HOME=&quot;$RESULTSDIR/.nextflow&quot;

if [[ -z &quot;$PATIENT&quot; || -z &quot;$CASENAME&quot; ]]; then
    echo &quot;Error: Could not extract PATIENT or CASENAME from the samplesheet.&quot;
    exit 1
fi


cd $RESULTSDIR

if [[ &quot;$GENOME&quot; == &quot;hg19&quot; ]]; then
    PROFILE=&quot;biowulf_test_run_slurm&quot;
elif [[ &quot;$GENOME&quot; == &quot;mm39&quot; ]]; then
    PROFILE=&quot;biowulf_mouse_RNA_slurm&quot;
else
    echo &quot;Unknown genome: $GENOME&quot;
    exit 1
fi

logname=$(basename &quot;$SAMPLESHEET&quot; .csv)
timestamp=$(date +&quot;%Y%m%d-%H%M%S&quot;)
sbatch &lt;&lt;EOT
#!/bin/bash
#SBATCH --job-name=&quot;$logname&quot;
#SBATCH --output=&quot;$OUTDIR/$PATIENT/$CASENAME/${logname}_%A_${timestamp}.out&quot;
#SBATCH --cpus-per-task=2
#SBATCH --mem=05g
#SBATCH --time=08-00:00:00

module load nextflow/23.10.0 singularity graphviz
nextflow run -c $CONFIG_FILE -profile $PROFILE --logdir $LOG $WF_HOME/main.nf -resume --samplesheet $SAMPLESHEET --resultsdir $OUTDIR --genome_v $GENOME --platform $PLATFORM
exit 0
EOT
</code></pre></div> </details> <p>Download or copy the script into your Biowulf working directory and make it executable:</p> <div class=highlight><pre><span></span><code>cp /data/khanlab/projects/Nextflow_dev/dev/AWS_POC_MVP_NF/launch.sh .
chmod +x launch.sh
</code></pre></div> <h3 id=step-2-view-usage-instructions>Step 2 — View usage instructions<a class=headerlink href=#step-2-view-usage-instructions title="Permanent link">&para;</a></h3> <p>Run the script without arguments to display the usage help:</p> <div class=highlight><pre><span></span><code>./launch.sh

Usage: ./launch.sh &lt;samplesheet_with_full_path&gt; [output_directory] [genome]
This script requires at least one positional argument:
1. Path to samplesheet
Optional arguments:
2. Path to results directory (default: /data/khanlab/projects/processed_DATA)
3. Genome name. Accepted values are hg19 and mm39 (default: hg19)
</code></pre></div> <p>Usage Examples:</p> <ol> <li> <p>Run with defaults (output to default path, genome hg19):</p> <p><code>./launch.sh /path/to/samplesheet.csv</code></p> </li> <li> <p>Run with custom output directory:</p> <p><code>./launch.sh /path/to/samplesheet.csv /custom/output/path</code></p> </li> <li> <p>Run with custom output directory and genome:</p> <p><code>./launch.sh /path/to/samplesheet.csv /custom/output/path mm39</code></p> </li> </ol> <div class="admonition tip"> <p class=admonition-title>mm39</p> <p>If mm39 is selected for the genome, the pipeline will run <em>Mapping &amp; Gene expression</em> only for mouse data.</p> </div> <h2 id=preparing-the-samplesheet>Preparing the Samplesheet<a class=headerlink href=#preparing-the-samplesheet title="Permanent link">&para;</a></h2> <p>You can prepare a samplesheet in two ways, depending on your use case:</p> <ol> <li><strong>Build from a Master Sheet</strong> – Recommended if you also plan to visualize results in the <a href=https://oncogenomics.ccr.cancer.gov/production/public/ >ClinOmics Data Portal</a>. </li> <li><strong>Build Manually</strong> – For standalone processing without portal integration.</li> </ol> <h3 id=1-build-from-a-master-sheet-recommended>1. Build from a Master Sheet (Recommended)<a class=headerlink href=#1-build-from-a-master-sheet-recommended title="Permanent link">&para;</a></h3> <p>For Khanlab workflows, master sheets are stored on Biowulf in:</p> <p><code>/data/khanlab/projects/DATA/Sequencing_Tracking_Master</code></p> <p>Use the script:</p> <p><a href=https://github.com/CCRGeneticsBranch/Oncogenomics_NF_WF/blob/feature/casename/samplesheet_builder.py>samplesheet_builder.py</a> <div class=highlight><pre><span></span><code>python ./samplesheet_builder.py &lt;PatientID&gt; &lt;CaseName&gt;
</code></pre></div></p> <p>Default Master Sheet Directory: /data/khanlab/projects/DATA/Sequencing_Tracking_Master</p> <p>Default Input Directory: /data/khanlab/projects/DATA</p> <p>To use custom directories, edit DEFAULT_SAMPLESHEET_DIR and DEFAULT_INPUT_DIR in the script.</p> <p><strong>Error Handling:</strong></p> <ol> <li><strong>Invalid FASTQ Paths</strong> – If <code>read1</code> or <code>read2</code> paths are invalid, the script will print an error prompting you to verify input paths.</li> <li><strong>Unknown Capture Kit</strong> – If the capture kit entered is not listed in <a href=../#sequencing-capture-kits>Sequencing Capture Kits</a>, the script will assign <code>unknown</code> as the kit name.</li> </ol> <h3 id=2-build-your-own-samplesheet>2. Build Your Own Samplesheet<a class=headerlink href=#2-build-your-own-samplesheet title="Permanent link">&para;</a></h3> <p>Create a CSV with these column information:</p> <table> <thead> <tr> <th>Column</th> <th>Description</th> <th>Example</th> </tr> </thead> <tbody> <tr> <td>sample</td> <td>Patient name</td> <td>NCI-Test1</td> </tr> <tr> <td>library</td> <td>Sample library name</td> <td>Test1_T1D_E</td> </tr> <tr> <td>read1</td> <td>Path to R1 FASTQ</td> <td>/data/khanlab/DATA/Sample_Test1_T1D_E/Sample_Test1_T1D_E.R1.fastq.gz</td> </tr> <tr> <td>read2</td> <td>Path to R2 FASTQ</td> <td>/data/khanlab/DATA/Sample_Test1_T1D_E/Sample_Test1_T1D_E.R2.fastq.gz</td> </tr> <tr> <td>sample_captures</td> <td>Capture kit name</td> <td><a href=../#sequencing-capture-kits>Sequencing Capture Kits</a></td> </tr> <tr> <td>Matched_RNA</td> <td>Matched RNA library (optional)</td> <td>Test1_T1R_T</td> </tr> <tr> <td>Matched_normal</td> <td>Matched exome normal library (optional)</td> <td>Test1_N1D_E</td> </tr> <tr> <td>Diagnosis</td> <td>Patient diagnosis</td> <td>Glioma</td> </tr> <tr> <td>casename</td> <td>Case name</td> <td>NCI-Test1</td> </tr> <tr> <td>type</td> <td>Data type</td> <td>tumor_RNA, tumor_DNA, normal_DNA, etc.</td> </tr> <tr> <td>FCID</td> <td>Flowcell ID (optional)</td> <td>ACJ678349</td> </tr> <tr> <td>Genome</td> <td>Genome</td> <td>hg19 or mm39</td> </tr> </tbody> </table> <p><strong>Example Input csv:</strong></p> <table> <thead> <tr> <th>sample</th> <th>library</th> <th>read1</th> <th>read2</th> <th>sample_captures</th> <th>Diagnosis</th> <th>Matched_RNA</th> <th>Matched_normal</th> <th>casename</th> <th>type</th> <th>FCID</th> <th>Genome</th> </tr> </thead> <tbody> <tr> <td>Test8</td> <td>Test5_T1D_E</td> <td>/data/khanlab/projects/fastq/Test5_T1D_E_R1.fastq.gz</td> <td>/data/khanlab/projects/fastq/Test5_T1D_E_R2.fastq.gz</td> <td>clin.ex.v1</td> <td>Osteosarcoma</td> <td></td> <td>Test8_N2D_E</td> <td>NFtest0523</td> <td>tumor_DNA</td> <td>AWXYNH2</td> <td>hg19</td> </tr> <tr> <td>Test8</td> <td>Test8_N2D_E</td> <td>/data/khanlab/projects/fastq/Test8_N2D_E_R1.fastq.gz</td> <td>/data/khanlab/projects/fastq/Test8_N2D_E_R2.fastq.gz</td> <td>clin.ex.v1</td> <td>Osteosarcoma</td> <td></td> <td></td> <td>NFtest0523</td> <td>normal_DNA</td> <td>AWXYNH2</td> <td>hg19</td> </tr> </tbody> </table> <aside class=md-source-file> <span class=md-source-file__fact> <span class=md-icon title="Last update"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M21 13.1c-.1 0-.3.1-.4.2l-1 1 2.1 2.1 1-1c.2-.2.2-.6 0-.8l-1.3-1.3c-.1-.1-.2-.2-.4-.2m-1.9 1.8-6.1 6V23h2.1l6.1-6.1-2.1-2M12.5 7v5.2l4 2.4-1 1L11 13V7h1.5M11 21.9c-5.1-.5-9-4.8-9-9.9C2 6.5 6.5 2 12 2c5.3 0 9.6 4.1 10 9.3-.3-.1-.6-.2-1-.2s-.7.1-1 .2C19.6 7.2 16.2 4 12 4c-4.4 0-8 3.6-8 8 0 4.1 3.1 7.5 7.1 7.9l-.1.2v1.8Z"/></svg> </span> 2025-08-14 </span> </aside> </article> </div> <script>var tabs=__md_get("__tabs");if(Array.isArray(tabs))e:for(var set of document.querySelectorAll(".tabbed-set")){var tab,labels=set.querySelector(".tabbed-labels");for(tab of tabs)for(var label of labels.getElementsByTagName("label"))if(label.innerText.trim()===tab){var input=document.getElementById(label.htmlFor);input.checked=!0;continue e}}</script> <script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script> </div> </main> <footer class=md-footer> <div class="md-footer-meta md-typeset"> <div class="md-footer-meta__inner md-grid"> <div class=md-copyright> <div class=md-copyright__highlight> Copyright &copy; 2025 CCR Genetics Branch </div> Made with <a href=https://squidfunk.github.io/mkdocs-material/ target=_blank rel=noopener> Material for MkDocs </a> </div> </div> </div> </footer> </div> <div class=md-dialog data-md-component=dialog> <div class="md-dialog__inner md-typeset"></div> </div> <script id=__config type=application/json>{"base": "..", "features": ["content.code.copy", "content.code.annotate", "content.tabs.link"], "search": "../assets/javascripts/workers/search.b8dbb3d2.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script> <script src=../assets/javascripts/bundle.bd41221c.min.js></script> </body> </html>